### PR TITLE
Docs: mirroring

### PR DESCRIPTION
**Pull request by ozkatz**
Add documentation to lakeFS Mirroring

### CODE REVIEW

**Review by idanovo — CHANGES_REQUESTED** [review]
Overall looks good!
Beside what I wrote, I think we have to add a clarification that this feature is currently supported only in s3, and also paragraph about GC(?).

File: docs/howto/mirroring.md [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+Typically, object stores provide a replication/batch copy API to allow for disaster recovery: as new objects are written, they are asynchronously copied to other geo locations. 
+
+In the case of regional failure, users can rely on the other geolocations which should contain relatively-up-to-date state.
+
+The problem is reasoning about what managed to arrive by the time of disaster and what hasn't:
+* have all the necessary files for a given dataset arrived?
+* in cases there are dependencies between datasets, are all dependencies also up to date?
+* what is currently in-flight or haven't even started replicating yet?
+
+Reasoning about these is non-trivial, especially in the face of a regional disaster, However ensuring business continuity might require that we have these answers.
```
Comment (itaiad200) [inline docs/howto/mirroring.md:42]:
```suggestion
Reasoning about these is non-trivial, especially in the face of a regional disaster, however ensuring business continuity might require that we have these answers.
```

File: docs/howto/mirroring.md [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+
+We can train our model in region A, and a month later feed the same commit ID into another region - and get back the same results.
+
+
+## Setting up mirroring
+
+### Configuring bucket replication on S3
+
+The objects within the repository are copied using your cloud provider's object store replication mechanism.
+For AWS S3, please refer to the [AWS S3 replication documentation](https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-how-setup.html) to make sure your lakeFS repository's storage namespace is replicated.
```
Comment (itaiad200) [inline docs/howto/mirroring.md:65]: "to make sure your lakeFS repository's storage namespace is replicated."
Not clear enough. Possible customers questions: What's a storage namespace? What path should be the target?

File: docs/howto/mirroring.md [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+Once a user has been created and the replication policy attached to it, create an access key and secret to be used by the mirroring process.
+
+
+### Authorizing the lakeFS Mirror process to use the replication user
+
+Please [contact Treeverse customer success](mailto:support@treeverse.io?subject=Setting up replication) to connect the newly created user with the mirroring process  
+
+
+### Configuring repository replication
```
Comment (itaiad200) [inline docs/howto/mirroring.md:179]: Do we want to expose all of this to potential customers at this point?

File: docs/howto/mirroring.md [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+-X POST \
+--data '{
+    "name": "<MIRROR_NAME>",
+    "region": "<MIRROR_REGION>",
+    "storage_namespace": "<MIRROR_STORAGE_NAMESPACE>"
+}'
+```
+Using the following parameters:
+
+* `ORGANIZATION_ID` - The ID as it appears in the URL of your lakeFS installation (e.g. `https://my-org.us-east-1.lakefscloud.io/`)
```
Comment (itaiad200) [inline docs/howto/mirroring.md:196]: That's the org name, isn't it?

File: docs/howto/mirroring.md [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+curl --location 'https://<ORGANIZATION_ID>.<SOURCE_REGION>.lakefscloud.io/service/replication/v1/repositories/<SOURCE_REPO>/mirrors' \
+--header 'Content-Type: application/json' \
+-u <ACCESS_KEY_ID>:<SECRET_ACCESS_KEY> \
+-X POST \
+--data '{
+    "name": "<MIRROR_NAME>",
+    "region": "<MIRROR_REGION>",
+    "storage_namespace": "<MIRROR_STORAGE_NAMESPACE>"
+}'
+```
```
Comment (itaiad200) [inline docs/howto/mirroring.md:193]: Example output or format

File: docs/howto/mirroring.md [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+
+## Other replication operations
+
+### Listing all mirrors for a repository
+
+```bash
+curl --location 'https://<ORGANIZATION_ID>.<SOURCE_REGION>.lakefscloud.io/service/replication/v1/repositories/<SOURCE_REPO>/mirrors' \
+-u <ACCESS_KEY_ID>:<SECRET_ACCESS_KEY> -s
+```
```
Comment (itaiad200) [inline docs/howto/mirroring.md:248]: Example output or format

File: docs/howto/mirroring.md [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+### Getting a specific mirror
+
+```bash
+url --location 'https://<ORGANIZATION_ID>.<SOURCE_REGION>.lakefscloud.io/service/replication/v1/repositories/<SOURCE_REPO>/mirrors/<MIRROR_ID>' \
+-u <ACCESS_KEY_ID>:<SECRET_ACCESS_KEY> -s
+```
```
Comment (itaiad200) [inline docs/howto/mirroring.md:255]: Example output or format

File: docs/howto/mirroring.md [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+### Deleting a specific mirror
+
+```bash
+curl --location --request DELETE 'https://<ORGANIZATION_ID>.<SOURCE_REGION>.lakefscloud.io/service/replication/v1/repositories/<SOURCE_REPO>/mirrors/<MIRROR_ID>' \
+-u <ACCESS_KEY_ID>:<SECRET_ACCESS_KEY>
+```
```
Comment (itaiad200) [inline docs/howto/mirroring.md:262]: Example output or format

File: docs/howto/mirroring.md [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+
+## Limitations
+
+1. Mirroring is currently only supported on [AWS S3](https://aws.amazon.com/s3/) and [lakeFS Cloud for AWS](https://lakefs.cloud)
+1. Read-only mirrors cannot be written to. Mirroring is one-way, from source to destination(s)
```
Comment (itaiad200) [inline docs/howto/mirroring.md:267]: A single mirror is indeed one-way. However I don't think it's clear enough that you can create multiple mirrors from any region to any region

File: docs/howto/mirroring.md [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+1. Currently, only branches are mirrored. Tags and arbitrary commits that do not belong to any branch are not replicated
+1. [lakeFS Hooks](./hooks) will only run on the source repository, not its replicas
+1. Replication is still asynchronous: reading from a branch will always return a valid commit that this branch has pointed to, but it is not guaranteed to be the **latest commit** this branch is pointing to.
```
Comment (itaiad200) [inline docs/howto/mirroring.md:270]:
```suggestion
1. Replication is still asynchronous: reading from a branch will always return a valid commit that the source branch has pointed to, but it is not guaranteed to be the **latest commit** the source branch is pointing to.
```

File: docs/howto/mirroring.md [hunk 11]
```diff
@@ -0,0 +1,270 @@
+---
+title: Mirroring
+parent: How-To
+description: Mirroring allows replicating commits between lakeFS installations in different geo-locations/regions
+---
+
+# Mirroring
```
Comment (itaiad200) [inline docs/howto/mirroring.md:7]: This should have BETA written all over it :)

**Review by itaiad200 — APPROVED** [review]
Thank you, it's very clear.

File: docs/howto/mirroring.md [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+curl --location --request DELETE 'https://<ORGANIZATION_ID>.<SOURCE_REGION>.lakefscloud.io/service/replication/v1/repositories/<SOURCE_REPO>/mirrors/<MIRROR_ID>' \
+-u <ACCESS_KEY_ID>:<SECRET_ACCESS_KEY>
+```
+
+## Limitations
+
+1. Mirroring is currently only supported on [AWS S3](https://aws.amazon.com/s3/) and [lakeFS Cloud for AWS](https://lakefs.cloud)
+1. Read-only mirrors cannot be written to. Mirroring is one-way, from source to destination(s)
+1. Currently, only branches are mirrored. Tags and arbitrary commits that do not belong to any branch are not replicated
+1. [lakeFS Hooks](./hooks) will only run on the source repository, not its replicas
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply docs/howto/mirroring.md:272]: we introduce read-only repositories here, they don't exist anywhere else.. and are limited :)

File: docs/howto/mirroring.md (hunk 5 above)
Comment (ozkatz) [reply docs/howto/mirroring.md:196]: I don't have such an example (taken from R&D notes..) - feel free to add a better reference in separate PR!

File: docs/howto/mirroring.md (hunk 11 above)
Comment (ozkatz) [reply docs/howto/mirroring.md:7]: Added in the title!

File: docs/howto/mirroring.md [hunk 13]
```diff
@@ -0,0 +1,270 @@
...
+            "fs:ReadCommit",
+            "fs:ListCommits",
+            "fs:CreateBranch",
+            "fs:DeleteBranch",
+            "fs:RevertBranch",
+            "fs:ReadBranch",
+            "fs:ListBranches"
+         ],
+         "effect": "allow",
+         "resource": "arn:lakefs:fs:::repository/{mirrorId}"
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply docs/howto/mirroring.md:161]: Taken from your .md :)

File: docs/howto/mirroring.md [hunk 14]
```diff
@@ -0,0 +1,270 @@
...
+Using lakeFS mirroring makes it much easier to answer: we are guaranteed that the latest commit that exists in the replica is in a consistent state and is fully usable, even if it isn't the absolute latest commit - it still reflects a known, consistent, point in time.
+
+
+### Data Locality
+
+For certain workloads, it might be cheaper to have data available in multiple regions: Expensive hardware such as GPUs might fluctuate in price, so we'd want to pick the region that currently offers the best pricing. The difference could easily offset to cost of the replicated data.
+
+The challenge is reproducibility - Say we have an ML training job that reads image files from a path in the object store. Which files existed at the time of training?
+
+If data is constantly flowing between regions, this might be harder to answer than we think. And even if we know - how can we recreate that exact state if we want to run the process again (for example, to rebuild that model for troubleshooting).
```
[the remark this answers is not in the document]
Comment (ozkatz) [reply docs/howto/mirroring.md:56]: Might rephrase in a separate PR later :)

**Review by idanovo — APPROVED** [review]
Thanks!

File: docs/howto/mirroring.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (guy-har) [reply docs/howto/mirroring.md:272]: That's true, I probably read it wrong, but it sounds to me like the limitation mentioned here is read-only repositories are read-only repositories. I would expect the limitation to be that mirrored repositories are read-only repositories and can not be written to.

**Pull request merged**